pkg/transparentproxy/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,119 @@
+	"strconv"
+
+	kube_core "k8s.io/api/core/v1"
+
+	"github.com/kumahq/kuma/pkg/plugins/runtime/k8s/metadata"
+	"github.com/kumahq/kuma/pkg/transparentproxy/config"
+)
+
+type PodRedirect struct {
+	BuiltinDNSEnabled     bool
+	BuiltinDNSPort        uint32
+	ExcludeOutboundPorts  string
+	RedirectPortOutbound  uint32
+	RedirectInbound       bool
+	ExcludeInboundPorts   string
+	RedirectPortInbound   uint32
+	RedirectPortInboundV6 uint32
+	UID                   string
+}
+
+func NewPodRedirectForPod(pod *kube_core.Pod) (*PodRedirect, error) {
+	var err error
+	podRedirect := &PodRedirect{}
+
+	podRedirect.BuiltinDNSEnabled, _, err = metadata.Annotations(pod.Annotations).GetEnabled(metadata.KumaBuiltinDNS)
+	if err != nil {
+		return nil, err
+	}
+
+	podRedirect.BuiltinDNSPort, _, err = metadata.Annotations(pod.Annotations).GetUint32(metadata.KumaBuiltinDNSPort)
+	if err != nil {
+		return nil, err
+	}
+
+	podRedirect.ExcludeOutboundPorts, _ = metadata.Annotations(pod.Annotations).GetString(metadata.KumaTrafficExcludeOutboundPorts)
+
+	podRedirect.RedirectPortOutbound, _, err = metadata.Annotations(pod.Annotations).GetUint32(metadata.KumaTransparentProxyingOutboundPortAnnotation)
+	if err != nil {
+		return nil, err
+	}
+
+	podRedirect.RedirectInbound = true
+	enabled, exist, err := metadata.Annotations(pod.Annotations).GetEnabled(metadata.KumaGatewayAnnotation)
+	if err != nil {
+		return nil, err
+	}
+	if exist && enabled {
+		podRedirect.RedirectInbound = false
+	}
+
+	podRedirect.ExcludeInboundPorts, _ = metadata.Annotations(pod.Annotations).GetString(metadata.KumaTrafficExcludeInboundPorts)
+
+	podRedirect.RedirectPortInbound, _, err = metadata.Annotations(pod.Annotations).GetUint32(metadata.KumaTransparentProxyingInboundPortAnnotation)
+	if err != nil {
+		return nil, err
+	}
+
+	podRedirect.RedirectPortInboundV6, _, err = metadata.Annotations(pod.Annotations).GetUint32(metadata.KumaTransparentProxyingInboundPortAnnotationV6)
+	if err != nil {
+		return nil, err
+	}
+
+	podRedirect.UID, _ = metadata.Annotations(pod.Annotations).GetString(metadata.KumaSidecarUID)
+
+	return podRedirect, nil
+}
+
+func (pr *PodRedirect) AsTransparentProxyConfig() *config.TransparentProxyConfig {
+	return &config.TransparentProxyConfig{
+		DryRun:                 false,
+		Verbose:                false,
+		RedirectPortOutBound:   fmt.Sprintf("%d", pr.RedirectPortOutbound),
+		RedirectInBound:        pr.RedirectInbound,
+		RedirectPortInBound:    fmt.Sprintf("%d", pr.RedirectPortInbound),
+		RedirectPortInBoundV6:  fmt.Sprintf("%d", pr.RedirectPortInboundV6),
+		ExcludeInboundPorts:    pr.ExcludeInboundPorts,
+		ExcludeOutboundPorts:   pr.ExcludeOutboundPorts,
+		UID:                    pr.UID,
+		GID:                    pr.UID, // TODO: shall we have a separate annotation here?
+		RedirectDNS:            pr.BuiltinDNSEnabled,
+		RedirectAllDNSTraffic:  false,
+		AgentDNSListenerPort:   fmt.Sprintf("%d", pr.BuiltinDNSPort),
+		DNSUpstreamTargetChain: "",
+	}
+}
```

Comment:
is this function some kind of middle step to do some other refactor? I don't see it being used anywhere

Reply:
This is going to be used in the CNI.